tests/adapter/dbt/tests/adapter/constraints/test_constraints.py

```diff
@@ -62,14 +97,91 @@ def test__constraints_wrong_column_names(self, project):
 
         assert contract_actual_config is True
 
-        expected_compile_error = "Please ensure the name, order, and number of columns in your `yml` file match the columns in your SQL file."
-        expected_schema_file_columns = "Schema File Columns: ['ID', 'COLOR', 'DATE_DAY']"
-        expected_sql_file_columns = "SQL File Columns: ['ERROR', 'COLOR', 'DATE_DAY']"
+        expected_compile_error = "Please ensure the name, data_type, order, and number of columns in your `yml` file match the columns in your SQL file."
+        expected_schema_file_columns = (
+            f"Schema File Columns: id {int_type}, color {string_type}, date_day DATE"
+        )
+        expected_sql_file_columns = (
+            f"SQL File Columns: error {int_type}, color {string_type}, date_day DATE"
+        )
 
         assert expected_compile_error in log_output
         assert expected_schema_file_columns in log_output
         assert expected_sql_file_columns in log_output
 
+    def test__constraints_wrong_column_data_types(
+        self, project, string_type, int_type, schema_int_type, data_types
+    ):
+        for (sql_column_value, schema_data_type, error_data_type) in data_types:
+            # Write parametrized data_type to sql file
+            write_file(
```

Comment:
It would be nice for this function to be a fixture instead, so that we can yield the file and then clean it up after. Maybe the `project` fixture does that if the file is setup in the project directory?

Reply:
The `project` fixture [drops the test schema on teardown](https://github.com/dbt-labs/dbt-core/blob/main/core/dbt/tests/fixtures/project.py#L519), but from what I can tell, it does not clean up the temp project directory. I understand this might actually be by design, for ease of debugging failing tests - [similar to intentionally preserving log files](https://github.com/dbt-labs/dbt-core/pull/6908#discussion_r1100677783). Worth revisiting, but probably out of scope for this PR.